pkg/ipam/node.go

```diff
@@ -546,8 +569,18 @@ func (n *Node) determineMaintenanceAction() (*maintenanceAction, error) {
 		return nil, err
 	}
 
+	surgeAllocate := 0
+	numPendingPods, err := getPendingPodCnt(n.name)
+	if err != nil {
+		scopedLog.WithError(err).Infof("Unable to compute pending pods, pod watcher might be initializing")
+	} else if numPendingPods > stats.NeededIPs {
+		surgeAllocate = numPendingPods - stats.NeededIPs
+	}
+
 	n.mutex.RLock()
-	a.allocation.MaxIPsToAllocate = stats.NeededIPs + n.getMaxAboveWatermark()
+	// handleIPAllocation() takes a min of MaxIPsToAllocate and IPs available for allocation on the interface.
+	// This makes sure we don't try to allocate more than what's available.
+	a.allocation.MaxIPsToAllocate = stats.NeededIPs + n.getMaxAboveWatermark() + surgeAllocate
```

Comment:
Might be worth documenting this new behavior in the ENI docs, specially here:

https://github.com/cilium/cilium/blob/284acfecd3398b65734915f73a6d3ce17ecce351/Documentation/concepts/networking/ipam/eni.rst#ip-allocation

I could also see us using `max(maxAboveWatermark, surgeAllocate)` instead of `maxAboveWatermark + surgeAllocate`. The latter is a bit more wasteful, but also more intuitive I guess. So I think using `+` is fine here.

Reply:
Makes sense. Updated the docs to mention surge allocation. Yeah, I decided on `+` instead of `max` because surge allocate would also eventually adhere with watermark defined by different parameters. Doing `+` would ensure we always satisfy `maxAboveWatermark` too.